pkg/controller/nodeipam/ipam/multi_cidr_range_allocator.go

```diff
@@ -0,0 +1,1158 @@
+			}
+		}
+	}
+
+	if isDualstack {
+		// In case of dualstack CIDRs, currently the default values for PerNodeMaskSize are
+		// 24 for IPv4 (PerNodeHostBits=8) and 64 for IPv6(PerNodeHostBits=64), there is no
+		// requirement for the PerNodeHostBits to be equal for IPv4 and IPv6, However with
+		// the introduction of ClusterCIDRs, we enforce the requirement for a single
+		// PerNodeHostBits field, thus we choose the minimum PerNodeHostBits value, to avoid
+		// overflow for IPv4 CIDRs.
+		if ipv4PerNodeHostBits >= minPerNodeHostBits && ipv4PerNodeHostBits <= ipv6PerNodeHostBits {
+			defaultCIDRConfig.Spec.PerNodeHostBits = ipv4PerNodeHostBits
+		} else if ipv6PerNodeHostBits >= minPerNodeHostBits && ipv6PerNodeHostBits <= ipv4MaxCIDRMask {
+			defaultCIDRConfig.Spec.PerNodeHostBits = ipv6PerNodeHostBits
+		}
+	}
+
+	existingConfigList.Items = append(existingConfigList.Items, *defaultCIDRConfig)
+
+	return
+}
+
+// reconcileCreate handles create ClusterCIDR events.
+func (r *multiCIDRRangeAllocator) reconcileCreate(clusterCIDR *networkingv1alpha1.ClusterCIDR) error {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	if needToAddFinalizer(clusterCIDR, clusterCIDRFinalizer) {
+		klog.V(3).Infof("Creating ClusterCIDR %s", clusterCIDR.Name)
+		if err := r.createClusterCIDR(clusterCIDR, false); err != nil {
+			klog.Errorf("Unable to create ClusterCIDR %s : %v", clusterCIDR.Name, err)
+			return err
+		}
+	}
+	return nil
+}
+
+// reconcileBootstrap handles creation of existing ClusterCIDRs.
+// adds a finalizer if not already present.
+func (r *multiCIDRRangeAllocator) reconcileBootstrap(clusterCIDR *networkingv1alpha1.ClusterCIDR) error {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	terminating := false
+	// Create the ClusterCIDR only if the Spec has not been modified.
+	if clusterCIDR.Generation > 1 {
+		terminating = true
+		err := fmt.Errorf("CIDRs from ClusterCIDR %s will not be used for allocation as it was modified", clusterCIDR.Name)
+		klog.Errorf("ClusterCIDR Modified: %v", err)
+	}
+
+	klog.V(2).Infof("Creating ClusterCIDR %s during bootstrap", clusterCIDR.Name)
+	if err := r.createClusterCIDR(clusterCIDR, terminating); err != nil {
+		klog.Errorf("Unable to create ClusterCIDR %s: %v", clusterCIDR.Name, err)
+		return err
+	}
+
+	return nil
+}
+
+// createClusterCIDR creates and maps the cidrSets in the cidrMap.
+func (r *multiCIDRRangeAllocator) createClusterCIDR(clusterCIDR *networkingv1alpha1.ClusterCIDR, terminating bool) error {
+	nodeSelector, err := r.nodeSelectorKey(clusterCIDR)
+	if err != nil {
+		return fmt.Errorf("unable to get labelSelector key: %w", err)
+	}
+
+	clusterCIDRSet, err := r.createClusterCIDRSet(clusterCIDR, terminating)
+	if err != nil {
+		return fmt.Errorf("invalid ClusterCIDR: %w", err)
+	}
+
+	if clusterCIDRSet.IPv4CIDRSet == nil && clusterCIDRSet.IPv6CIDRSet == nil {
+		return errors.New("invalid ClusterCIDR: must provide IPv4 and/or IPv6 config")
```

Comment:
nit, I think this is a panic or Fatalf, the ClusterCIDR has to be validated, hence there has to be always an IPv4 or IPv6

Reply:
The ClusterCIDR has been validated during the API validation: https://github.com/sarveshr7/kubernetes/blob/de6ef0052a352262dc2ca872344b8d41148cb5ea/pkg/apis/networking/validation/validation.go#L625-L628

This is validating the CIDRSet in which we are maintaining the CIDR allocation status

Reply:
I see, they can drift, is ok , thanks